### ISSUE TITLE
Unable to start server with more than 1 broker

### ISSUE THREAD

**Issue by tbashor** [SocketCluster/socketcluster#117]
When I set brokers in server.js to anything more than 1, I get the following error:
```
/usr/local/bin/node server.js
   [Busy] Launching SocketCluster
   >> Broker PID: 80776
   >> Broker ID: 0
   >> Broker PID: 80777
   >> Broker ID: 1
1445121659533 - Origin: Broker (PID 80777)
    [Error] Error: listen EADDRINUSE
    at exports._errnoException (util.js:746:11)
    at Server._listen2 (net.js:1139:19)
    at listen (net.js:1182:10)
    at Server.listen (net.js:1261:5)
    at EventEmitter.self.listen (/Users/developer/socketcluster-test/node_modules/socketcluster/node_modules/iocluster/node_modules/ndata/node_modules/ncom/index.js:139:19)
    at Object.<anonymous> (/Users/developer/socketcluster-test/node_modules/socketcluster/node_modules/iocluster/node_modules/ndata/server.js:480:10)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    !! Failed to bind to a port because it was already used by another process.
```

For example, I've tried with both workers and brokers set to 2 and with workers set to 3 and brokers set to 2. 

Any idea what could be wrong?

**Comment by jondubois** [comment]
@tbashor Does running as root fix the issue?

**Comment by jondubois** [comment]
Or try deleting the /tmp/socketcluster/ directory.

**Comment by jondubois** [comment]
What are your configs for the SocketCluster() instance in server.js?

**Comment by tbashor** [comment]
Starting it with sudo worked. Thanks!

Why is it necessary run as root? Is there a way to get it to work without the elevated access?

**Comment by jondubois** [comment]
@tbashor You shouldn't have to run as root. I just wanted to check what the problem was.
SC creates a /tmp/socketcluster/ directory to store unix domain socket file descriptors.

Maybe you had existing socket fds from an earlier time that you ran SC as root? Then when you try to run it as non-root, it tries to access the fds under the same directory but it can't because it doesn't have the privilege. Does deleting the /tmp/socketcluster directory allow you to run as non-root?

I may have to make this case more obvious in the error message because it's confusing.

**Comment by tbashor** [comment]
Deleting /tmp/socketcluster/ did not allow me to start the processes. Which is strange to me since when starting with only one broker does work but does not result in a new tmp/socket cluster directory. I have a /tmp/ directory in my user account but that does not have a socketcluster directory. 

The existing error is certainly misleading so it would be great if that was addressed. 

As far as the fix goes, how can I avoid running the process with sudo? Do I just need to change the user is related to the process after it starts?

**Comment by SomeKittens** [comment]
Also encountered this problem, running as sudo fixed it, there was no entry in `/tmp/socketcluster` before running as sudo.

OSX 10.10.5, if that helps.

**Comment by jondubois** [comment]
@tbashor What OS were you using?

**Comment by tbashor** [comment]
OSX 10.11.1

**Comment by SomeKittens** [comment]
Tried running the server with `sudo` (which created the `/tmp/socketcluster`) and chown'ing that directory to my local user, still getting EADDRINUSE.

**Comment by SomeKittens** [comment]
Running with brokers > 1 works fine with non-sudo on my Ubuntu 14.04.3 server, looks like this is an OSX thing.

Confirmed the error is also on OSX 10.11

**Comment by jondubois** [comment]
@SomeKittens @tbashor It's difficult for me to debug this issue because I'm running Linux.
You could try logging the output of ```self._getBrokerSocketPaths()``` in the SocketCluster code here: https://github.com/SocketCluster/socketcluster/blob/master/index.js#L552

When running with 2 brokers, you should expect to get an array like:
```js
[ '/tmp/socketcluster/350a68c6-b1ec-411b-a916-eb0cd6e067d2_c6d951f854/b0',
  '/tmp/socketcluster/350a68c6-b1ec-411b-a916-eb0cd6e067d2_c6d951f854/b1' ]
```

The important part is that you should have on ending with ```/b0``` and the other one ending with ```/b1```.

If that looks normal for you, you can have a look inside the ```iocluster``` submodule of SocketCluster on this line: https://github.com/SocketCluster/iocluster/blob/master/index.js#L308

^ That line should run twice (once for each broker) and ```socketPath``` should be a diferent path each time.

Then finally, if that still looks normal, you can dive deeper in the ```ndata``` submodule (of ```iocluster```).
You can log the ```SOCKET_PATH``` here: https://github.com/SocketCluster/ndata/blob/master/server.js#L480

This is where the broker server actually tries to bind itself to the UNIX domain socket and likely where the error happens.

**Comment by SomeKittens** [comment]
@tbashor If it helps at all, remember that `/tmp` on OSX is symlinked to `/private/tmp`.  Wonder if that's a contributor.

@jondubois I'll check into that when I get a break later today.  Side note: How many brokers should we be running?  Looking at the docs, they're for syncing across machines, so no more than one per box?

**Comment by jondubois** [comment]
@SomeKittens Often, one broker is enough, but it depends on how heavily you use channels within your system - Every use case is different.

The main purpose of brokers to allow workers to share data between workers (which are running on different CPU cores) - They do this automatically. You can also add logic to sync them across multiple machines.

**Comment by happilymarrieddad** [comment]
I'm not sure if this helps but one of our production servers has only 1 broker and it is handling a LOT of tasks (4 worker cores and 6 remote socketcluster-clients along with sometimes over 100 users accessing the system). Despite the number of tasks, the cpu never really goes above 2%. 1 broker should be enough I would think for almost any number of tasks. We are using AWS EC2 if that helps at all.

**Comment by jondubois** [comment]
This has been fixed so that it now works with multiple brokers on OSX without having to use sudo. The fix is in socketcluster v5.0.2 .
